### PR TITLE
Add support for tracestate extraction from OpenTelemetry API

### PR DESCRIPTION
**Pull request by PerfectSlayer**
# What Does This Do

This PR will ensure that extracted context from OpenTelemetry API will respect W3C tracestate if any.
Such context can then be used as parent span or span links.

# Motivation

This will improve OpenTelemetry interoperability.

# Additional Notes

Previously, the W3C tracestate was dropped and not propagated to any downstream service.

Jira ticket: [APMJAVA-1060]

[APMJAVA-1060]: https://datadoghq.atlassian.net/browse/APMJAVA-1060?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: dd-java-agent/instrumentation/opentelemetry/opentelemetry-1.4/src/main/java/datadog/trace/instrumentation/opentelemetry14/context/propagation/TraceStateHelper.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.lang.reflect.Method;
+import java.util.function.Function;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class helps to decode W3C tracestate header into a {@link TraceState} instance using OTel
+ * API.
```
Comment (mcculls) [inline dd-java-agent/instrumentation/opentelemetry/opentelemetry-1.4/src/main/java/datadog/trace/instrumentation/opentelemetry14/context/propagation/TraceStateHelper.java:15]: Given the encoding follows a public spec, is there a benefit of using this internal class (which could be moved in a future release) over our previous implementation of the same spec?

We need to provide a compatible fallback anyway, for situations where we can't find the encoder/decoder so why not just use that?

**Review by mcculls — APPROVED** [review]
LGTM - only question I had was whether it was worth trying to use `internal.W3CTraceContextEncoding` vs just using our implementation of the W3C encoding spec

File: dd-java-agent/instrumentation/opentelemetry/opentelemetry-1.4/src/main/java/datadog/trace/instrumentation/opentelemetry14/context/propagation/TraceStateHelper.java (hunk 1 above)
Comment (PerfectSlayer) [reply dd-java-agent/instrumentation/opentelemetry/opentelemetry-1.4/src/main/java/datadog/trace/instrumentation/opentelemetry14/context/propagation/TraceStateHelper.java:15]: Meaning we will vendor OTel tracestate encoding/decoding implementation? Doing so, we won’t have their fixes anymore.

I  originally also went for call OTel implementation to make sure we do have the same behavior.
I remembered @bantonsson had a hard time with W3C style propagation implementation.

Comment (mcculls) [reply dd-java-agent/instrumentation/opentelemetry/opentelemetry-1.4/src/main/java/datadog/trace/instrumentation/opentelemetry14/context/propagation/TraceStateHelper.java:15]: Unless the spec is evolving then I'd expect there to be very few (if any) changes to the code - and we already have to provide a fall-back implementation. The approach in the latest `W3CTraceContextEncoding` looks straightforward

Comment (bantonsson) [reply dd-java-agent/instrumentation/opentelemetry/opentelemetry-1.4/src/main/java/datadog/trace/instrumentation/opentelemetry14/context/propagation/TraceStateHelper.java:15]: I'm a bit at a loss about what we are talking about here, but the difficulties with the `W3CHttpCodec` and pulling out the `PropagationTags` from the W3C `TraceState` header is rooted in doing that performant, without materializing a `Map` with `String` 
 parts for all the vendor states et.c.

**Pull request merged**